### PR TITLE
Drop note is public 2159

### PR DESCRIPTION
**Pull request by oldPadavan**
I want to merge this change because...

implements #2159

### Pull Request Checklist

1. [x] Privileged views and APIs are guarded by proper permission checks.
1. [x] All visible strings are translated with proper context.
1. [x] All data-formatting is locale-aware (dates, numbers, and so on).
1. [x] Database queries are optimized and the number of queries is constant.
1. [x] The changes are tested.
1. [x] The code is documented (docstrings, project documentation).

### CODE REVIEW

**Comment by oldPadavan** [comment]
Although in #2159 it was mentioned the notes should be visible to staff only, I left visible to the user its own notes. If it's not ok, I'll modify the view

File: saleor/dashboard/order/forms.py [hunk 1]
```diff
@@ -248,14 +248,10 @@ def save(self, commit=True):
...
         widgets = {
             'content': forms.Textarea()}
-        labels = {
-            'content': pgettext_lazy('Order note', 'Note'),
-            'is_public': pgettext_lazy(
-                'Allow customers to see note toggle',
-                'Customer can see this note')}
+        labels = {'content': pgettext_lazy('Order note', 'Note')}
 
     def send_confirmation_email(self):
```
Comment (Pacu2) [inline saleor/dashboard/order/forms.py]: I think this should be removed as well, was used for notifying an user about new notes on his order

File: saleor/order/models.py [hunk 2]
```diff
@@ -322,6 +322,7 @@ class Meta:
 class OrderNote(BaseNote):
     order = models.ForeignKey(
         Order, related_name='notes', on_delete=models.CASCADE)
+    is_public = None
```
Comment (Pacu2) [inline saleor/order/models.py]: Why is that needed?

File: saleor/dashboard/order/forms.py (hunk 1 above)
Comment (Pacu2) [inline saleor/dashboard/order/forms.py]: I think that was used for notifying a customer about a new note on his order, could be removed as well

File: saleor/order/models.py (hunk 2 above)
Comment (oldPadavan) [reply saleor/order/models.py]: `OrderNote` inherits from Abstract base class `BaseNote`. In order to remote the field `is_public` from `OrderNote` I set it to None. From docs:
> Fields inherited from abstract base classes can be overridden with another field or value, or be removed with None.

File: saleor/dashboard/order/forms.py (hunk 1 above)
Comment (oldPadavan) [reply saleor/dashboard/order/forms.py]: Thanks for comments, I missed it

File: saleor/order/views.py [hunk 3]
```diff
@@ -31,10 +31,10 @@ def details(request, token):
     orders = orders.select_related(
         'billing_address', 'shipping_address', 'user')
     order = get_object_or_404(orders, token=token)
-    notes = order.notes.filter(is_public=True)
+    user = request.user if request.user.is_authenticated else None
+    notes = order.notes.filter(user=user)
```
Comment (Pacu2) [inline saleor/order/views.py]: I see some trouble here, let's say if staff user added some note, and his account was deleted after, then `OrderNote` he added will be preserved, but `user` will be set to `None`. This may cause some data leakage. 

I wonder if we shouldn't leave `is_public` on the `OrderNote` to determine if the note was added by the customer or not, without a possibility to create a public note from the dashboard 

Or shall we leave `OrderNote` to customers only, and introduce another way for staff to add notes about orders/ fulfillments etc. 
cc @maarcingebala @NyanKiyoshi

Comment (maarcingebala) [reply saleor/order/views.py]: I'd remove this line at all and disable rendering notes in the customer's view. As in the original issue - those notes should be rewritten so that they're only avaialble to staff users.

To keep the customer notes functionality, I would add a simple Textfield in Cart and in Order models. We would use that to store a one-time note from the customer and display it somewhere in the dashboard.

Comment (oldPadavan) [reply saleor/order/views.py]: Thanks for Your feedback! I have a couple of questions. Should these new notes be visible to the customer? Should these fields be added in this PR or it should be a separate PR?

Comment (maarcingebala) [reply saleor/order/views.py]: If users can add notes to their order I think they should also be able to see it later in the storefront's order details page. I would prefer to do those changes in this PR to keep the "order notes" funtionality available on `master`.

**Comment by maarcingebala** [comment]
@oldPadavan A few remarks about this PR:
- I think we could entirely drop the `BaseNote` abstract model. It was a missed concept and it won't be used anywhere else except `OrderNote`, so we could simply move its fields to `OrderNote`.
- It would be great if you could remove those "merge" commits. We prefer to use `git rebase` to keep up with `master` as it produces cleaner commit history.

**Comment by maarcingebala** [comment]
Btw. I just noticed that we've already merged 7 PRs from you. We really appreciate your contribution! I'd like to reach you outside of Github, could you please drop me an email to `marcin@mirumee.com` when you have time and I'll write you back? Thanks 🙂

**Comment by oldPadavan** [comment]
I noticed there's also `CustomerNote` model which inherits from `BaseNote`

**Comment by maarcingebala** [comment]
You're right - we also use it to add multiple notes in the customer details page. Personally, I think it's overcomplicated now because there is also a simple text field that allows defining customer notes there. But let's keep it then, I don't want to change too much in one PR.

File: saleor/checkout/utils.py [hunk 4]
```diff
@@ -828,7 +828,8 @@ def _fill_order_with_cart_data(order, cart, discounts, taxes):
             order, line.variant, line.quantity, discounts, taxes)
 
     if cart.note:
-        order.notes.create(user=order.user, content=cart.note)
+        order.customer_note = cart.note
+        order.save()
```
Comment (Pacu2) [inline saleor/checkout/utils.py]: if saving just one field, we could use `updated_fields` here

File: saleor/order/migrations/0051_order_customer_note.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
+# Generated by Django 2.0.3 on 2018-07-31 14:05
```
Comment (Pacu2) [inline saleor/order/migrations/0051_order_customer_note.py]: I think we could squash this migration with the previous one

File: tests/test_order.py [hunk 6]
```diff
@@ -443,3 +433,24 @@ def test_create_user_after_order(order, client):
     user = User.objects.filter(email='hello@mirumee.com').first()
     assert user is not None
     assert user.orders.filter(token=order.token).exists()
+
+
+def test_view_order_detals(order, client):
```
Comment (Pacu2) [inline tests/test_order.py]: Shouldn't it be `details` instead of `detals`?

File: saleor/order/forms.py [hunk 7]
```diff
@@ -49,11 +49,11 @@ def __init__(self, *args, **kwargs):
         self.fields['email'].widget = forms.HiddenInput()
 
 
-class OrderNoteForm(forms.ModelForm):
+class CustomerNoteForm(forms.ModelForm):
+    customer_note = forms.CharField(
+        max_length=250, required=False, strip=True, label=False,
```
Comment (Pacu2) [inline saleor/order/forms.py]: Aren't we missing a label here?

Comment (oldPadavan) [reply saleor/order/forms.py]: There was no label in note form during checkout so I didn't add it here as well. I can add a label or simply a header like "Add a note to Your order" or leave it as it is. How would You advice is better?

File: saleor/checkout/utils.py (hunk 4 above)
Comment (NyanKiyoshi) [reply saleor/checkout/utils.py]: We could pass this field data at creation time instead:
https://github.com/mirumee/saleor/blob/a945504c5fc1e1d7471212681761751864e8773e/saleor/checkout/utils.py#L828-L836

File: saleor/order/forms.py (hunk 7 above)
Comment (NyanKiyoshi) [reply saleor/order/forms.py]: The current one was "Add note to order", it seems good enough.

File: saleor/order/models.py [hunk 8]
```diff
@@ -90,6 +90,7 @@ class Order(models.Model):
         decimal_places=settings.DEFAULT_DECIMAL_PLACES, default=0)
     discount_name = models.CharField(max_length=255, default='', blank=True)
     display_gross_prices = models.BooleanField(default=True)
+    customer_note = models.TextField(blank=True, default='')
```
Comment (NyanKiyoshi) [inline saleor/order/models.py]: Isn't a `TextField` a bit huge?

File: saleor/order/forms.py [hunk 9]
```diff
@@ -49,11 +49,11 @@ def __init__(self, *args, **kwargs):
         self.fields['email'].widget = forms.HiddenInput()
 
 
-class OrderNoteForm(forms.ModelForm):
+class CustomerNoteForm(forms.ModelForm):
```
Comment (NyanKiyoshi) [inline saleor/order/forms.py]: I think the class should be named `OrderNoteForm` instead of `CustomerNoteForm`.

File: saleor/dashboard/order/forms.py [hunk 10]
```diff
@@ -248,18 +247,10 @@ def save(self, commit=True):
 class OrderNoteForm(forms.ModelForm):
     class Meta:
         model = OrderNote
-        fields = ['content', 'is_public']
+        fields = ['content']
```
Comment (NyanKiyoshi) [inline saleor/dashboard/order/forms.py]: Do we still want to use a one to many relation for the dashboard note form?

File: saleor/order/views.py [hunk 11]
```diff
@@ -31,12 +31,9 @@ def details(request, token):
...
-    notes = order.notes.filter(is_public=True)
-    ctx = {'order': order, 'notes': notes}
-    if order.is_open():
-        user = request.user if request.user.is_authenticated else None
-        note = OrderNote(order=order, user=user)
-        note_form = OrderNoteForm(request.POST or None, instance=note)
+    ctx = {'order': order}
+    if order.is_open() and not order.customer_note:
+        note_form = CustomerNoteForm(request.POST or None, instance=order)
         ctx.update({'note_form': note_form})
```
Comment (NyanKiyoshi) [inline saleor/order/views.py]: What about we set `note_form` to `None` and then, we move `ctx = {...}` at the end and add everything we need to the context to save some useless instructions?

```python
def details(request, token):
    note_form = None
    orders = Order.objects.confirmed().prefetch_related(
        'lines__variant', 'fulfillments', 'fulfillments__lines',
        'fulfillments__lines__order_line')
    orders = orders.select_related(
        'billing_address', 'shipping_address', 'user')
    order = get_object_or_404(orders, token=token)
    if order.is_open() and not order.customer_note:
        note_form = CustomerNoteForm(request.POST or None, instance=order)
        ctx.update({'note_form': note_form})
        if request.method == 'POST':
            if note_form.is_valid():
                note_form.save()
                return redirect('order:details', token=order.token)
    fulfillments = order.fulfillments.filter(
        status=FulfillmentStatus.FULFILLED)
    ctx = {
        'order': order, 'fulfillments': fulfillments, 'note_form': note_form}
    return TemplateResponse(request, 'order/details.html', ctx)

File: templates/dashboard/order/detail.html [hunk 12]
```diff
@@ -518,6 +518,22 @@ <h5 class="data-table-title">
...
+      <div class="card order-notes{% if not order.customer_note %} print-hide{% endif %}" id="customer-note">
+        <div class="card-content">
+          <span class="card-title black-text">
+            {% trans "Customer note" context "Customer order note card title" %}
+          </span>
+          {% if order.customer_note %}
+            <p>{{ order.customer_note }}</p>
+          {% else %}
+            <p>
+              {% trans "There's no note from customer associated with this order." context "No customer order note card message" %}
```
Comment (NyanKiyoshi) [inline templates/dashboard/order/detail.html]: Suggested change:
> There **is** no note from **the** customer associated with this order.

Or maybe we could just hide the whole card?

File: templates/order/details.html [hunk 13]
```diff
@@ -63,35 +63,11 @@ <h2 class="order-details__header">
...
-            {% else %}
-              {% blocktrans with date=note.date %}
-                {{ date }} by anonymous customer
-              {% endblocktrans %}
-            {% endif %}
-          </small>
-          <p>{{ note.content }}</p>
-        {% endfor %}
+      {% if order.customer_note %}
+        <h2 class="order-notes__header">{% trans "Customer note" context "Customer order note" %}</h2>
```
Comment (NyanKiyoshi) [inline templates/order/details.html]: I would say "Your notes" instead since it is in the customer's panel.

File: saleor/order/models.py (hunk 8 above)
Comment (oldPadavan) [reply saleor/order/models.py]: `Cart` also contains `note` which is later copied to `Order.customer_note` (previously was added to the list of `OrderNote`). That field is `TextField`, so I decided to keep it the same in `Order`.

File: templates/order/details.html (hunk 13 above)
Comment (oldPadavan) [reply templates/order/details.html]: Sounds better, just I would rename to "Your note" as it could be at most 1 note

File: saleor/order/forms.py (hunk 7 above)
Comment (oldPadavan) [reply saleor/order/forms.py]: I've added a header to the template

File: saleor/dashboard/order/forms.py (hunk 10 above)
Comment (oldPadavan) [reply saleor/dashboard/order/forms.py]: Do You mean to leave only one-time comment here as well?

Comment (maarcingebala) [reply saleor/dashboard/order/forms.py]: This one should use a relation - admins are able to add multiple notes to an order.
